e2e/fuzz/replay/replay_message_reader.go

```diff
@@ -135,28 +160,59 @@ func (r *replayMessageReader) startChunkReading(messagesChannel chan []*ReplayMe
 	}()
 }
 
-// checks if given node is done with execution of messages
-func (r *replayMessageReader) checkIfDoneWithExecution(validatorId pbft.NodeID, msg *pbft.MessageReq) {
-	//when nodes reach sequence that is higher than the sequence in file,
-	//or if it is in constant round change we know we either reached the end of execution or nodes are stuck and there is a problem
-	if msg.View.Sequence > r.lastSequence ||
-		isSequenceInContinuousRoundChange(msg, validatorId) {
+// processMessage checks if message is a timeout, or if all messages have been read from the queue
+func (r *replayMessageReader) processMessage(validatorId pbft.NodeID, msg *pbft.MessageReq) bool {
+	if isTimeoutMessage(msg) {
+		return true
+	}
+
+	if msg.View.Sequence > r.lastSequenceMessages[validatorId].sequence ||
+		(msg.View.Sequence == r.lastSequenceMessages[validatorId].sequence && r.areMessagesFromLastSequenceProcessed(msg, validatorId)) {
 		r.lock.Lock()
-		if _, isDone := r.nodesDoneWithExecution[validatorId]; !isDone {
+		_, isDone := r.nodesDoneWithExecution[validatorId]
+		if !isDone {
 			r.nodesDoneWithExecution[validatorId] = true
 			r.msgProcessingDone <- string(validatorId)
 		}
 		r.lock.Unlock()
 	}
+
+	return false
+}
+
+// areMessagesFromLastSequenceProcessed checks if all the messages from the last sequence of given node are processed so that the node can be stoped
+func (r *replayMessageReader) areMessagesFromLastSequenceProcessed(msg *pbft.MessageReq, validatorId pbft.NodeID) bool {
+	lastSequenceMessages := r.lastSequenceMessages[validatorId]
+
+	if len(lastSequenceMessages.messages) == 0 {
+		return true
+	} else {
+		messageIndexToRemove := -1
+		for i, message := range lastSequenceMessages.messages {
+			if areMessagesEqual(msg, message) {
+				messageIndexToRemove = i
+			}
+		}
+
+		if messageIndexToRemove != -1 {
+			lastSequenceMessages.messages = append(lastSequenceMessages.messages[:messageIndexToRemove], lastSequenceMessages.messages[messageIndexToRemove+1:]...)
+		}
+
+		return len(lastSequenceMessages.messages) == 0
+	}
```

Comment:
Small suggestion in order to spare a line of code or so :slightly_smiling_face: 

```suggestion
        lastSeqMsgsCount := len(lastSequenceMessages.messages)
	if lastSeqMsgsCount > 0 {
		messageIndexToRemove := -1
		for i, message := range lastSequenceMessages.messages {
			if areMessagesEqual(msg, message) {
				messageIndexToRemove = i
			}
		}

		if messageIndexToRemove != -1 {
			lastSequenceMessages.messages = append(lastSequenceMessages.messages[:messageIndexToRemove], lastSequenceMessages.messages[messageIndexToRemove+1:]...)
		}
	}
        return lastSeqMsgsCount == 0
```

Reply:
since we change the count of messages in if messageIndexToRemove != -1, we can not use the lastSeqMsgsCount variable like this. But will use something from suggestion.